### PR TITLE
Fixed #28694 -- Made django.utils.text.slugify() strip dashes and underscores.

### PR DESCRIPTION
**Pull request by smithdc1**
[Ticket #28694](https://code.djangoproject.com/ticket/28694)
Previous work on this ticket at #8733

I've added an additional test for the documented behaviour of removing duplicate dashes and spaces betweeen words. I've left it here as a separate commit - happy to raise as a separate PR if you would prefer. 

I was a little unsure if to add to the release notes. I went on the side of caution as I figgured it's eaiser to delete something. 😄

### CODE REVIEW

File: django/utils/text.py [hunk 1]
```diff
@@ -402,7 +402,7 @@ def slugify(value, allow_unicode=False):
         value = unicodedata.normalize('NFKC', value)
     else:
         value = unicodedata.normalize('NFKD', value).encode('ascii', 'ignore').decode('ascii')
-    value = re.sub(r'[^\w\s-]', '', value.lower()).strip()
+    value = re.sub(r'[^\w\s-]', '', value.lower()).strip('-_ ')
```
Comment (felixxm) [inline django/utils/text.py:405]: `strip()` removes also other whitespace, e.g. `\t`. I think we should move strip to the next line, i.e.
```python
value = re.sub(r'[^\w\s-]', '', value.lower())
return re.sub(r'[-\s]+', '-', value).strip('-_')
```

File: tests/utils_tests/test_text.py [hunk 2]
```diff
@@ -192,6 +192,11 @@ def test_slugify(self):
             # given - expected - Unicode?
             ('Hello, World!', 'hello-world', False),
             ('spam & eggs', 'spam-eggs', False),
+            (' multiple---dash and  space ', 'multiple-dash-and-space', False),
+            ('underscore_in-value', 'underscore_in-value', False),
+            ('__strip__underscore-value___', 'strip__underscore-value', False),
+            ('--strip-dash-value---', 'strip-dash-value', False),
+            ('__strip-mixed-value---', 'strip-mixed-value', False),
```
Comment (felixxm) [inline tests/utils_tests/test_text.py:199]: Can we add tests also for mixed cases?, e.g. `'_ -value _-'`

**Comment by felixxm** [comment]
Please update also `slugify()`'s docstring.

File: docs/ref/utils.txt [hunk 3]
```diff
@@ -867,6 +867,10 @@ appropriate entities.
         >>> slugify('你好 World', allow_unicode=True)
         '你好-world'
 
+    .. versionchanged:: 3.2
+
+        This function now also removes trailing underscores and dashes.
```
Comment (felixxm) [inline docs/ref/utils.txt:872]: `In older versions, leading and trailing underscores and dashes weren't removed.`

File: django/utils/text.py (hunk 1 above)
Comment (smithdc1) [reply django/utils/text.py:405]: Just had a thought, should we strip all punctuation?

Comment (felixxm) [reply django/utils/text.py:405]: We've already removed them in `re.sub(r'[^\w\s-]', '', value.lower())`.

**Comment by felixxm** [comment]
@smithdc1 Thanks :+1:

**Pull request merged**